### PR TITLE
Make xformers optional even if it is available

### PR DESCRIPTION
**Pull request by kn**
Problem:
xformers being available doesn't always mean it should be used. For example, I install xformers because it makes inference fast but I find xformers make my dreambooth training slower.

Solution:
Require explicit flag to enable xformers when running training scripts.

Related recent change: https://github.com/huggingface/diffusers/pull/1640

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
cc @patil-suraj @pcuenca, I tend to agree here with @kn . If we make it fully optional we should however not wrap it into a "try .... else ..." statement, but just let it fail if it's not available

File: examples/dreambooth/train_dreambooth.py [hunk 1]
```diff
@@ -509,14 +510,17 @@ def main(args):
...
+        if is_xformers_available():
+            try:
+                unet.enable_xformers_memory_efficient_attention()
+            except Exception as e:
+                logger.warning(
+                    "Could not enable memory efficient attention. Make sure xformers is installed"
+                    f" correctly and a GPU is available: {e}"
+                )
+        else:
+            logger.warning("Could not enable memory efficient attention. Make sure xformers is installed correctly.")
```
Comment (pcuenca) [inline examples/dreambooth/train_dreambooth.py:523]:
```suggestion
    if args.use_xformers:
            unet.enable_xformers_memory_efficient_attention()
```

I agree with @patrickvonplaten here, if the user explicilty requests they want `xformers`, let's fail if it's not available (it's important that they know it's not working in this case).

File: examples/text_to_image/train_text_to_image.py [hunk 2]
```diff
@@ -383,14 +384,17 @@ def main():
...
+        if is_xformers_available():
+            try:
+                unet.enable_xformers_memory_efficient_attention()
+            except Exception as e:
+                logger.warning(
+                    "Could not enable memory efficient attention. Make sure xformers is installed"
+                    f" correctly and a GPU is available: {e}"
+                )
+        else:
+            logger.warning("Could not enable memory efficient attention. Make sure xformers is installed correctly.")
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:397]:
```suggestion
    if args.use_xformers:
            unet.enable_xformers_memory_efficient_attention()
```

Same as above

File: examples/textual_inversion/textual_inversion.py [hunk 3]
```diff
@@ -458,14 +459,17 @@ def main():
...
+        if is_xformers_available():
+            try:
+                unet.enable_xformers_memory_efficient_attention()
+            except Exception as e:
+                logger.warning(
+                    "Could not enable memory efficient attention. Make sure xformers is installed"
+                    f" correctly and a GPU is available: {e}"
+                )
+        else:
+            logger.warning("Could not enable memory efficient attention. Make sure xformers is installed correctly.")
```
Comment (pcuenca) [inline examples/textual_inversion/textual_inversion.py:472]:
```suggestion
    if args.use_xformers:
            unet.enable_xformers_memory_efficient_attention()
```

Same as above

**Review by pcuenca — APPROVED** [review]
Thanks, I think this makes sense!

**Comment by kn** [comment]
good idea on raising an exception when xformers is used but not available! PTAL

**Comment by patrickvonplaten** [comment]
I'm in favor of this as it better follows the "no optimizations by default" API of PyTorch and should be "less surprising" for the user long-term. 

This change might suddenly slow down some notebooks, so also kindly pinging @apolinario and @patil-suraj here.

File: examples/dreambooth/train_dreambooth.py [hunk 4]
```diff
@@ -241,6 +241,7 @@ def parse_args(input_args=None):
         ),
     )
     parser.add_argument("--local_rank", type=int, default=-1, help="For distributed training: local_rank")
+    parser.add_argument("--use_xformers", action="store_true", help="Whether or not to use xformers.")
```
Comment (patil-suraj) [inline examples/dreambooth/train_dreambooth.py:244]: could call this arg `enable_xformers_memory_efficient_attention` to align with the method name ?

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for the PR, agree with it, makes sense to add an explicit argument here to enable exformers. Just left a nit. Also, let's add a note  in the examples readme about this new argument.

**Comment by kn** [comment]
renamed the flag and added a note in READMEs - PTAL

**Review by patil-suraj — APPROVED** [review]
Thanks a lot!

**Comment by patrickvonplaten** [comment]
Thanks for the PR @kn !

**Pull request merged**